File: java/dsa/algorithms/sorting/frequency-sort.java
public String frequencySort(String s) {
    Map<Character, Integer> counts = new HashMap<>();
    for (char c : s.toCharArray()) {
        counts.put(c, counts.getOrDefault(c, 0) + 1);
    }
    
    int n = s.length();
    List<Character>[] buckets = new List[n + 1];
    for (char c : counts.keySet()) {
        int freq = counts.get(c);
        if (buckets[freq] == null) buckets[freq] = new ArrayList<>();
        buckets[freq].add(c);
    }
    
    StringBuilder res = new StringBuilder();
    for (int f = n; f >= 1; f--) {
        if (buckets[f] != null) {
            for (char c : buckets[f]) {
                for (int i = 0; i < f; i++) res.append(c);
            }
        }
    }
    return res.toString();
}

File: python/dsa/algorithms/sorting/frequency-sort.py
def frequencySort(s):
    from collections import Counter
    counts = Counter(s)
    
    n = len(s)
    buckets = [[] for _ in range(n + 1)]
    for char, freq in counts.items():
        buckets[freq].append(char)
        
    res = []
    for f in range(n, 0, -1):
        for char in buckets[f]:
            res.append(char * f)
            
    return "".join(res)